### PR TITLE
General opam 2 fixes!

### PR DESCRIPTION
**Pull request by dra27**
Logic error inadvertently introduced in #256 - `apt-get update` isn't necessarily called before the OCaml packages are installed.

Presently testing it in https://travis-ci.org/dra27/merlin/builds/454589966 which is where I found the issue...

### CODE REVIEW

**Comment by dra27** [comment]
Added 3 further alterations:
 - `OPAM_VERSION` was not respected by #255 - it now is, as is `OPAM_VERSION=2` or `OPAM_VERSION=2.0` (as for the Docker update in #270). macOS installs opam via Homebrew, so a warning is displayed if `OPAM_VERSION` is not compatible with the version in homebrew.
 - If the opam file is from version 1, then Warning 57 will be triggered by opam 2.0.1 or later when linting - this is selectively disabled. This pulls in a dependency on `opam-file-format`
 - Finally, and unrelatedly, build system switched from Jbuilder to Dune...

File: .travis-ocaml.sh [hunk 1]
```diff
@@ -18,17 +18,34 @@ set -uex
 # the ocaml version to test
 OCAML_VERSION=${OCAML_VERSION:-latest}
 SYS_OCAML_VERSION=4.02
-OPAM_VERSION=${OPAM_VERSION:-2.0.0}
+# Default opam is the latest release of opam 2
+OPAM_VERSION=${OPAM_VERSION:-2}
 OPAM_INIT=${OPAM_INIT:-true}
 
+case $OPAM_VERSION in
+    2|2.0) OPAM_VERSION=2.0.1;;
```
Comment (dra27) [inline .travis-ocaml.sh]: The idea is that this matrix will want updating as new releases of opam are made.

File: .travis-ocaml.sh [hunk 2]
```diff
@@ -18,17 +18,34 @@ set -uex
...
+esac
+
+if [ "$TRAVIS_OS_NAME" = "osx" ] ; then
+    brew update &> /dev/null
+    BREW_OPAM_VERSION=$(brew info opam --json=v1 | sed -e 's/.*"versions":{[^}]*"stable":"//' -e 's/".*//')
```
Comment (dra27) [inline .travis-ocaml.sh]: Somewhat evil, but I didn't want to pull (back) in a dependency on `jq` as well

**Comment by dra27** [comment]
It seems a mistake that the CI for this repo always uses the `.travis-opam.sh` from the `master` branch, rather than the one from PR (last commit)?

**Comment by samoht** [comment]
All green, thanks for the fix!

File: .travis.yml [hunk 3]
```diff
@@ -1,6 +1,5 @@
 language: c
 sudo: required
-install: wget https://raw.githubusercontent.com/ocaml/ocaml-ci-scripts/master/.travis-opam.sh
```
Comment (liyishuai) [inline .travis.yml]: Removing this line breaks the [instructions](/ocaml/ocaml-ci-scripts/blob/master/README-travis.md#opam-package-travis-opamsh):
> 2. Copy [.travis.yml](https://raw.githubusercontent.com/ocaml/ocaml-ci-scripts/master/.travis.yml) at the root of your project.

Comment (dra27) [reply .travis.yml]: Hmm - perhaps `test -e .travis.opam || wget ...`

**Pull request merged**